### PR TITLE
Fixes #8290 - dry up foreman_url logic in one place

### PR DESCRIPTION
**Pull request by dustints**
depends on https://github.com/theforeman/foreman/pull/1902

### CODE REVIEW

**Comment by ares** [comment]
Looks good, could you please add unit tests for `Foreman::Controller::ForemanUrlRenderable`?

File: app/controllers/concerns/foreman/controller/foreman_url_renderable.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    if @template_url && @host.try(:token).present?
+      foreman_url_from_uri(action, @template_url, host.token)
+    elsif proxy.present? && proxy.try(:features).map(&:name).include?('Templates') && host.try(:token).present?
+      foreman_url_from_templates_smart_proxy_plugin(action, proxy, host.token)
+    else
+      foreman_url_from_config_or_request(action, host.token)
+    end
+  end
+
```
Comment (ares) [inline app/controllers/concerns/foreman/controller/foreman_url_renderable.rb:18]: everything below this line should be private methods I think

**Comment by ares** [comment]
Also could you please rebase the PR, #1902 was merged so you don't have to include the commit here (it causes conflicts)

**Comment by ares** [comment]
Meanwhile I tested the patch, works nicely (in both config template rendering and TFTP file rendering).

**Comment by dustints** [comment]
@ares thanks for testing and reviewing.
Updated
- rebased
- added unit test for foreman_url_renderable, added in #8291 but forgot to add it here.
- made most methods private in foreman_url_renderable

**Comment by dustints** [comment]
mysql builds failing because of jenkins config issue

```
14:37:43 ERROR: Publisher hudson.tasks.junit.JUnitResultArchiver aborted due to exception
14:37:43 hudson.AbortException: No test report files were found. Configuration error?
```

**Comment by orrabin** [comment]
there is a known mysql error: http://projects.theforeman.org/issues/8537
if that's the only error ignore it for now and test again after the issue is fixed

**Comment by orrabin** [comment]
[test]

**Comment by orrabin** [comment]
@dustints can you please rebase?

**Pull request closed without merging**